File: ClassEval_T/cpp/solution/SQLGenerator.cpp

#include <sstream>
#include <algorithm>

#include <string>
#include <vector>
#include <map>

class SQLGenerator {
public:
    SQLGenerator(const std::string& table_name);

    std::string select(const std::vector<std::string>& fields = {}, const std::string& condition = "") const;
    std::string insert(const std::map<std::string, std::string>& data) const;
    std::string update(const std::map<std::string, std::string>& data, const std::string& condition) const;
    std::string delete_query(const std::string& condition) const;
    std::string select_female_under_age(int age) const;
    std::string select_by_age_range(int min_age, int max_age) const;

private:
    std::string table_name;
};


SQLGenerator::SQLGenerator(const std::string& table_name) : table_name(table_name) {}


std::string SQLGenerator::select(const std::vector<std::string>& fields, const std::string& condition) const {
    std::string fields_str = "*";
    if (!fields.empty()) {
        fields_str = fields[0];
        for (size_t i = 1; i < fields.size(); ++i) {
            fields_str += ", " + fields[i];
        }
    }

    std::string sql = "SELECT " + fields_str + " FROM " + table_name;
    if (!condition.empty()) {
        sql += " WHERE " + condition;
    }
    return sql + ";";
}


std::string SQLGenerator::insert(const std::map<std::string, std::string>& data) const {
    std::ostringstream fields_stream, values_stream;

    for (auto it = data.begin(); it != data.end(); ++it) {
        if (it != data.begin()) {
            fields_stream << ", ";
            values_stream << ", ";
        }
        fields_stream << it->first;
        values_stream << "'" << it->second << "'";
    }

    std::string sql = "INSERT INTO " + table_name + " (" + fields_stream.str() + ") VALUES (" + values_stream.str() + ")";
    return sql + ";";
}


std::string SQLGenerator::update(const std::map<std::string, std::string>& data, const std::string& condition) const {
    std::ostringstream set_clause_stream;

    for (auto it = data.begin(); it != data.end(); ++it) {
        if (it != data.begin()) {
            set_clause_stream << ", ";
        }
        set_clause_stream << it->first << " = '" << it->second << "'";
    }

    std::string sql = "UPDATE " + table_name + " SET " + set_clause_stream.str();
    if (!condition.empty()) {
        sql += " WHERE " + condition;
    }
    return sql + ";";
}


std::string SQLGenerator::delete_query(const std::string& condition) const {
    std::string sql = "DELETE FROM " + table_name;
    if (!condition.empty()) {
        sql += " WHERE " + condition;
    }
    return sql + ";";
}


std::string SQLGenerator::select_female_under_age(int age) const {
    std::string condition = "age < " + std::to_string(age) + " AND gender = 'female'";
    return select({}, condition);
}


std::string SQLGenerator::select_by_age_range(int min_age, int max_age) const {
    std::string condition = "age BETWEEN " + std::to_string(min_age) + " AND " + std::to_string(max_age);
    return select({}, condition);
}

File: BatchTestTool/java/java automated tester/Test/to_be_tested/SQLGenerator.java
package to_be_tested;

import java.util.HashMap;
import java.util.Map;
import java.util.ArrayList;
import java.util.List;

public class SQLGenerator {

    public String tableName;

    public SQLGenerator(String tableName) {
        this.tableName = tableName;
    }

    public String insert(Map<String, String> data) {
        String fields = String.join(", ", data.keySet().stream().toArray(String[]::new));
        String values = String.join(", ", data.values().stream().map(Object::toString).toArray(String[]::new));
        String sql = "INSERT INTO " + tableName + " (" + fields + ") VALUES (" + values + ")";
        return sql + ";";
    }

    public String update(Map<String, String> data, String condition) {
        StringBuilder setClause = new StringBuilder();
        for (Map.Entry<String, String> entry : data.entrySet()) {
            setClause.append(entry.getKey()).append(" = '").append(entry.getValue()).append("', ");
        }
        String setClauseStr = setClause.toString().trim().replaceAll(", $", "");
        String sql = "UPDATE " + tableName + " SET " + setClauseStr + " WHERE " + condition;
        return sql + ";";
    }

    public String select(String[] fields, String condition) {
        if (fields == null) {
            fields = new String[] { "*" };
        } else {
            StringBuilder sb = new StringBuilder();
            for (String field : fields) {
                sb.append(field);
                sb.append(", ");
            }
            fields = sb.toString().split(", ");
        }
        StringBuilder sql = new StringBuilder("SELECT ");
        for (String field : fields) {
            sql.append(field);
            sql.append(" ");
        }
        sql.append("FROM ");
        sql.append(tableName);
        if (condition != null) {
            sql.append(" WHERE ");
            sql.append(condition);
        }
        return sql.toString() + ";";
    }

    public String selectFemaleUnderAge(int age) {
        return select(new String[] { "*" }, "age < " + age + " AND gender = 'female'");
    }

    public String selectByAgeRange(int minAge, int maxAge) {
        return select(new String[] { "*" }, "age BETWEEN " + minAge + " AND " + maxAge);
    }

    public String delete(String condition) {
        return "DELETE FROM " + tableName + " WHERE " + condition + ";";
    }
}
